### ISSUE TITLE
User RapidJSON to parse JSON input

### ISSUE THREAD

**Issue by mitzimorris** [stan-dev/cmdstan#719]
#### Summary:

work done for https://github.com/stan-dev/stan/pull/2793 would be better if it were done instead directly in CmdStan

#### Description:

CmdStan is the only Stan client which needs a JSON parser.

 - add directory `io/json` to `cmdstan/src/cmdstan` and put all files under `stan/src/stan/io/json` in that repo.

- add corresponding unit tests

- add directory `cmdstan/lib` and add ujson libs there

- update `cmdstan/src/command.hpp` accordingly.

- find models and datasets appropriate for testing memory usage and speed.  test methods `optimize` (typically used for very big models where sampling is too slow) and method `fixed_param` for models which do no estimation whatsoever - simply test speed of reading in very large 2D and 3D matrices.

#### Expected Output:

CmdStan should behave exactly as before, only parsing of JSON inputs should be a *lot* 
faster.

#### Additional Information:
Provide any additional information here.

#### Current Version:
v2.20.0

**Comment by mitzimorris** [comment]
@rok-cesnovar - do you want to keep working on this as part of CmdStan?  else I will.

**Comment by rok-cesnovar** [comment]
Yes, of course. Thank you for creating the issue. We need to sort out the open questions below. After that I will open a PR.

Also tagging @bob-carpenter for the answers to his questions/comments from [here](https://github.com/stan-dev/stan/pull/2793#issuecomment-518229120). Hopefully it wont be too confusing if I answer the questions here.

> This is presumably a test for handling conjugate character encoding in UTF-8.
> 
> What we want to do about this depends on whether we want to check that Stan files are encoded in legal UTF-8.
> 
> Stan requires the program itself to be written in ASCII, but suggests UTF-8 may be used for the content of comments as those will not be read.

Since we will not be parsing comments do we need to check for valid UTF-8 there? RapidJSON supposedly does validate utf-8 but for some reason those two cases did not trigger an error. Other tests for invalid UTF8 characters did however. Will try a bit harder to understand what is going on there and why.

> I'm not sure the exact boundary conditions matter much here.  If Stan's looking for an integer and gets a double value, it'll still raise an error.  It doesn't seem unreasonable for the parser to try to make sense of
> 1234567890123456789012345678901234567890 by representing it as a double.

Should we leave the default behaviour than? The default behaviour is that 1234567890123456789012345678901234567890 will get parsed as a double and will be passed to the `data_handler` as a double, which will then trigger an error in the data handler I presume. 

> Did you modify their code or how it's used?  Is there a way to do this without modifying their API?

I had to modify their parser code to throw an error if it tries to parse an integer to double (basically replicate the existing parser).

> We tried to be careful about boundary conditions and throw errors, but I'm not sure that's the right behavior for floating point.  What if someone gives us a 30-digit number like 0.123456789012345678901234567890?  We'll probably just throw away the unused digits.  Arguably, rounding to 0 does the same thing.

Agreed. Should we just switch that back then?

**Comment by rok-cesnovar** [comment]
So to re-iterate what are the open questions (for the first 2 I have solutions ready whatever we decide to do):

- is it fine if we round to zero if the parsed number is too small to represent in double? Existing parser threw an error in this case.

- if an integer is too large, is it fine if we parse it to double and let the data handler throw an error? Existing parser threw an error in this cases.

- do we need to validate UTF-8 in the comments? In general the RapidJSON parser does validate UTF-8 but there are some slight differences to the existing parser that need to be explored in case we need to validate UTF-8 in the comments (see above and previous PR thread).

@mitzimorris do you have any opinions on these?

**Comment by bob-carpenter** [comment]
> On Aug 9, 2019, at 6:21 AM, Rok Češnovar <notifications@github.com> wrote:
> 
> So to re-iterate what are the open questions (for the first 2 I have solutions ready whatever we decide to do):
> 
> 	• is it fine if we round to zero if the parsed number is too small to represent in double? Existing parser threw an error in this case.

OK by me.

> 
> 	• if an integer is too large, is it fine if we parse it to double and let the data handler throw an error? Existing parser threw an error in this cases.

I'd rather throw here.

We're going to move Stan to 64-bit integers at some point and we'll have to change the cutoff point if we convert.

I don't think converting to double is the end of the world here.


> 	• do we need to validate UTF-8 in the comments? In general the RapidJSON parser does validate UTF-8 but there are some slight differences to the existing parser that need to be explored in case we need to validate UTF-8 in the comments (see above and previous PR thread).

I don't think it's critical to validate character encodings.

Current behavior is that bytes within comments (after "//" or "#" or between "/*  */") are ignored.

Checking UTF-8 is more a matter of display than anything else.


> @mitzimorris do you have any opinions on these?

Please do collect other opinions.  Mine aren't meant to be definitive here.

**Comment by rok-cesnovar** [comment]
Ok, based on the discussion with @mitzimorris at Stancon, I decided to move this along and we can address my questions on the final PR review. 

I opened 2 PRs that will move the existing parser code to Cmdstan (https://github.com/stan-dev/cmdstan/pull/726) and remove it from Stan (https://github.com/stan-dev/stan/pull/2803). Once that is merged we can add RapidJSON to cmdstan. I think that makes the PRs cleaner and easier to review and the git history will also be easier to track in case of any unforeseen issues with RapidJSON. I am not expecting them, but if there were, it would make reverting very difficult with changes across 2 repos.
